### DISCUSSION TITLE
[Babel-template] Support for (dot) property accessors after TS's non null assertion (!)

### DISCUSSION THREAD

**Discussion by Arooba-git** [babel/babel#15220] [Q&A] [answered]
Hi 👋 
I'm using babel-template to generate an AST, where I keep getting 'Unexpected token' whenever it encounters access to object property after TS's post-fix non-null assertion:

<img width="880" alt="Screen Shot 2022-11-22 at 6 11 29 AM" src="https://user-images.githubusercontent.com/56495631/203685523-86f8879e-67b1-453b-a51d-56b97508073f.png">

Does `babel-template` not support (dot) property accessors after TS's non-null assertion or not support non-null assertion at all, or am I missing something?  🤔
TIA :)

**Comment by JLHwung** [comment]
I can't reproduce your issue on Babel REPL: https://babel.dev/repl#?browsers=&build=&builtIns=false&corejs=3.21&spec=false&loose=true&code_lz=IYQgdARkA&debug=false&forceAllTransforms=false&shippedProposals=false&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=module&lineWrap=true&presets=env%2Ctypescript&prettier=false&targets=&version=7.20.4&externalPlugins=&assumptions=%7B%7D

Since you are using the Babel parser, did you enable the `typescript` plugin?

**Reply by Arooba-git** [reply]
Thank you for the prompt response!.. Yes, here is my config: 
<img width="839" alt="Screen Shot 2022-11-24 at 12 31 21 PM" src="https://user-images.githubusercontent.com/56495631/203688033-e3e69031-7472-4b4f-b55b-57f27864c92b.png">

Everything else, except the above issue, works fine...

**Reply by liuxingbaoyu** [reply]
Can you provide a repo that reproduces it?

**Reply by Arooba-git** [reply]
> Can you provide a repo that reproduces it?

Sure.. this is the repo I am parsing and here is one of the files at which the error occurs to be exact:
https://github.com/angular/components/blob/main/src/material/menu/menu-trigger.ts

Actually, I have updated my post. Parser doesn't throw the error when parsing, `babel-template` throws it.
In this file (link above), specifically at line 653 you have this:
```
this.menu._animationDone
            .pipe(take(1), delay(0, asapScheduler), takeUntil(this._parentMaterialMenu!._hovered()))
            .subscribe(() => this.openMenu());
            
```

In my program, I am taking the argument of the `takeUntil` operator and attaching a `next()` call to it:
```
   if (!observableCompleted) {
        unsubscribeStatement = template.ast(`
            ${collector.name}.next();
            ${collector.name}.complete();
        `);
    }
```
    
This code works fine with all other node types but when my program tries to do it for `this._parentMaterialMenu!._hovered()` (and similar others with TS non-null assertion operator) , it throws the error shown in the post.

**Reply by Arooba-git** [reply]
And here is the full stack trace in case it might help :) 

<img width="921" alt="Screen Shot 2022-11-24 at 5 50 28 PM" src="https://user-images.githubusercontent.com/56495631/203736093-429cbf44-1ec3-4cb1-808b-277cc10732c4.png">

**Reply by Arooba-git** [reply]
Line 62 referred to in the trace, is this one in my code (calling the `babel/core`'s `transformFromAST `function):

<img width="656" alt="Screen Shot 2022-11-24 at 5 55 03 PM" src="https://user-images.githubusercontent.com/56495631/203737122-37073fb0-6166-4ce8-ae6b-92b9c54782f3.png">

**Reply by liuxingbaoyu** [reply]
https://babeljs.io/docs/en/babel-template#string-usage
Can you try this?

**Reply by Arooba-git** [reply]
Sure, let me try and get back to you!

**Reply by Arooba-git** [reply]
Ah, it worked with the syntactic placeholders !
<img width="414" alt="Screen Shot 2022-11-24 at 6 11 16 PM" src="https://user-images.githubusercontent.com/56495631/203743478-a66f57e5-96a7-402d-ab8a-b41cae57b742.png">

<img width="537" alt="Screen Shot 2022-11-24 at 6 10 31 PM" src="https://user-images.githubusercontent.com/56495631/203743808-3d2eece8-34ee-41ce-abbe-3ee20f6b613c.png">


Thanks! I didn't give them a try 'cause I assumed they would error out as well 😅

**Reply by Arooba-git** [reply]
btw any insight into why `.ast()` wouldn't work well with these types of nodes? 🤔

**Reply by liuxingbaoyu** [reply]
I think you are using https://babeljs.io/docs/en/babel-template#string-usage instead of https://babeljs.io/docs/en/babel-template#template-literal-usage.

```
template.ast`code`
```
```
template.ast(`code`)
```

**Reply by liuxingbaoyu** [reply]
Only passing in an ast node directly can work normally, and passing in the text code of the node will cause the parsing to fail.

**Reply by Arooba-git** [reply]
> Only passing in an ast node directly can work normally, and passing in the text code of the node will cause the parsing to fail.

You mean passing node as placeholder would still work in `.ast()`? But the doc explicitly says _if no placeholders are used_ 🤔 
<img width="769" alt="Screen Shot 2022-11-24 at 6 27 06 PM" src="https://user-images.githubusercontent.com/56495631/203747274-9b99aebd-eea0-43e8-b58a-540dcdf6921a.png">

**Reply by liuxingbaoyu** [reply]
It should work, but requires reparsing the incoming nodes, which will fail since the `typescript` plugin is not enabled by default. Whereas template strings allow direct manipulation of AST nodes, which does not require the plugin to be enabled.
https://babeljs.io/docs/en/babel-template#api
This would allow you to splice code text directly, but obviously the method you're using is better.

This has nothing to do with placeholders, but a template call in js.
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals
```
template.ast`code`
```
```
template.ast(`code`)
```

**Reply by Arooba-git** [reply]
I see.. thanks, I understand the issue now 👍 ... was focusing on the wrong part 😅.

**Reply by liuxingbaoyu** [reply]
![image](https://user-images.githubusercontent.com/30521560/203749481-690b006e-2127-4f95-b386-b268a1c4daed.png)
```const fn = (...args) => {console.log(args)};fn(`${{}}`);fn`${{}}`;```

Only 
```
template.ast`code${nodeObj}`
``` 
can get the node object.

**Reply by Arooba-git** [reply]
This made everything so much clearer now. Thank you for your time! 🙌

**Comment by JLHwung** [comment, marked as answer]
Since the question is now updated by using `@babel/template`, you can specify the plugin option in the second argument of the `template.ast` call:

```js
// template.ast`foo!.bar` will not work
template.ast(`foo!.bar`, { plugins: ["typescript"] })
```

**Reply by Arooba-git** [reply]
> // template.ast`foo!.bar` will not work

Hmmm strange.. because I tried this:
<img width="358" alt="Screen Shot 2022-11-25 at 6 23 54 AM" src="https://user-images.githubusercontent.com/56495631/203864952-d71e759b-ca1b-4af9-90c2-043031672bdb.png">

and it's still working fine with TS non-null assertion:

<img width="537" alt="Screen Shot 2022-11-24 at 6 10 31 PM" src="https://user-images.githubusercontent.com/56495631/203865164-9858a648-67d6-4c2c-b366-a6b23ccdc299.png">

as @liuxingbaoyu mentioned.. the issue was with using string format ` .ast(``) ` instead of template `.ast`` `(that I'm now using)

**Reply by liuxingbaoyu** [reply]
This works when the special node (!.) is given as an argument. And @JLHwung provides a method when using templates to parse code text containing special nodes.
```
template.ast`code.aaa.${nodeObj}`
```
vs
```
template.ast`code!.aaa.${nodeObj}`
```
For the second usage you need to pass in a second parameter to enable the plugin.

**Reply by Arooba-git** [reply]
Right, so I'm understanding that in my code, where I'm using  `${collector.name}.next();` above,  the `${collector.name}` resolves to a string `pluginmaterial!.hovered()`, but since it is given as a `template argument ` (${})  it doesn't need TS plugin to be specially enabled. But had it been passed directly **as text** as in  ``.ast`pluginmaterial!.hovered()``, I would have to manually enable the TS plugin. Is the correct?

**Reply by liuxingbaoyu** [reply]
Yes.
> the ${collector.name} resolves to a string pluginmaterial!.hovered()

 In addition, directly using the template as a parameter, you can directly pass in the node object, which avoids parsing. Only parsing needs to explicitly enable the plugin.

**Reply by Arooba-git** [reply]
Right, because of course the only way to _directly pass in the node object_ in a template string, is via javascript's template placeholders ${}.
Means JS's template strings, when used directly with a function name/expression (foo``), aka tagged templates, avoid/bypass reparsing anything passed in their placeholders ${}. 
Is that right? :)

**Reply by liuxingbaoyu** [reply]
Yes.

**Comment by Arooba-git** [comment]
And for anyone reading this thread in future, I had actually gotten confused by the usage of the term 'placeholders' in `babel-template `docs. I was mixing javascript's template placeholders (${}) with `babel-template`'s (identifier/syntactic) placeholders which are **two different things** 🤦‍♀️ . In my screenshot above, where the `babel-template` docs tell us _if no placeholders are used, just strings, you can use `.ast()`_, it was **not** actually referring to javascript's template placeholder ${}, but babel-template's (syntactic/identifier) placeholders:

<img width="398" alt="Screen Shot 2022-11-25 at 8 00 23 AM" src="https://user-images.githubusercontent.com/56495631/203872795-1b0fec71-a458-4098-a684-5536b73747dd.png">
